### ISSUE TITLE
[Doc]: gitwash deleting main branch

### ISSUE THREAD

**Issue by rcomer** [matplotlib/matplotlib#25857]
### Documentation Link

https://matplotlib.org/devdocs/devel/development_workflow.html#workflow-summary

### Problem

Our dev guide has advice to consider deleting your `main` branch, with a link to [this pretty old page](https://matthew-brett.github.io/pydagogue/gh_delete_master.html).  I see two problems with the linked advice

* It describes `master` instead of `main`, and it may not be obvious for someone who is new to open source that that is just an old name for the same thing.  Also there were good EDI reasons why `master` was dropped.

* the instructions for deleting the branch and setting up `placeholder` as your new default branch are fairly involved.  These days you can just rename your branch within GitHub, which I think will have the same practical effect with less effort.

### Suggested improvement

I see two options

1. just remove the advice about deleting the `main` branch (but retain the advice to not develop with it).

2. remove the advice about deleting the branch, but add some advice to rename it instead.  I think this would fit better in the [setting up](https://matplotlib.org/devdocs/devel/development_setup.html) instructions as you’d want to do it before you clone.

I have a slight preference for 2, but I think that’s just because my default branch is `placeholder`, having followed gitwash in the past.  I’m not sure I have a strong argument that it’s better to rename than to leave it as `main`.

**Comment by timhoffm** [comment]
I don't understand what the benefit of renaming is. local `main`, `upstream/main`, `origin/main` are still the distributed variants of the same branch. Renaming part of that feel confusing. IMHO one should either delete a branch or leave it in place (untouched).

**Comment by jklymak** [comment]
I think the only distinction is didactic.  When you start with git it's not really clear what the difference between main, upstream/main, origin/main all are and how they talk to each other.  When I started I thought they all did talk to each other somehow.  It's very elegant, but confusing at first, particularly if you are just trying to get something done.  Conversely if you never have a "main" it's clear that your "placeholder" is a completely separate thing and you need to sync it up manually.

**Comment by rcomer** [comment]
Having an out-of-date local `main` (or `placeholder`) shouldn’t matter if you are creating feature branches off `upstream/main`.

**Comment by timhoffm** [comment]
If we want to go into the remote+branch details, https://matplotlib.org/stable/devel/development_workflow.html should get one (or two) visualizations.

Something similar to (taken from https://www.tomasbeuzen.com/post/git-fork-branch-pull/):
![grafik](https://github.com/matplotlib/matplotlib/assets/2836374/0ecf3e1e-56d8-4860-a075-edf5b68b2efb)

This is still not exactly how I would draw it because it mixes setup (fork, clone) and the PR workflow (pull, push, pull-request). These should better be in separate diagrams. Possibly one would also want to annotate the used branch names.

**Comment by rcomer** [comment]
I think we should be cautious about adding content to the dev guide.  Nothing wrong with the above suggestions in themselves, but the dev guide as a whole is quite extensive.  I think there's a danger that puts people off reading it at all.  I once decided not to bother putting a small change in a script because the working practices document for the repo it was in was huge!

**Comment by oscargus** [comment]
I think it often makes sense to have a TLDR approach. Brief information and then you can read up on parts where needed. Also, there is a difference between "this what we expect you to do" and "this is how you do what we expect you to do and some help to get you there".

With that said, it is extremely hard to write text that is both flexible in the sense that one know what can be skipped, but also contains information for all different types of background knowledge...

**Comment by jklymak** [comment]
> I think we should be cautious about adding content to the dev guide.

I agree.  I think for git workflows we should state what flow people need to use (because it actually is different from personal repo use), and link out to a few guides we like that explain that workflow.  Gitwash started out as such a guide, and if a modern good one doesn't exist, maybe we need to write one, but I'm not sure it belongs on our site.

**Comment by matthew-brett** [comment]
Only to say - as the original author of Gitwash - that I'm totally happy for anyone to either - rewrite Gitwash itself - or start from scratch and suggest that we all use that instead of Gitwash.

**Comment by melissawm** [comment]
As a note - my initial suggestion was to leverage existing documentation from git/github (which has gotten a lot better) instead of writing our own. That would also reduce maintainer workload and ensure we would have it up-to-date. My approach here would be to go with the minimal information needed for basic development and link out to official docs for other, more complicated workflows. 

For example: https://docs.github.com/en/get-started/getting-started-with-git/about-remote-repositories

These also have the advantage that they are translated into a number of human languages, not just english.

**Comment by tacaswell** [comment]
I am very much in favor of replacing as much of our "how to git" docs with links to github's (or git's) documentation.  Compared to a decade ago, git and github/gitlab are orders of magnitude more mainstream than they were and we should not have the expectation of needing to document the basics.

We do need to keep a couple of special-to-us-things (like we like re basing more than many projects (but much less than others I know of)), but how to push/fetch, merge, commit, etc should just be links....
